bindings/matrix-sdk-ffi/src/room.rs

```diff
@@ -840,6 +840,97 @@ impl Room {
+            }
+        })));
+
+        Ok(handle)
+    }
+}
+
+impl From<matrix_sdk::room::request_to_join::RequestToJoinRoom> for RequestToJoin {
+    fn from(request: matrix_sdk::room::request_to_join::RequestToJoinRoom) -> Self {
+        Self {
+            event_id: request.event_id.to_string(),
+            user_id: request.member_info.user_id.to_string(),
+            room_id: request.room_id().to_string(),
+            display_name: request.member_info.display_name.clone(),
+            avatar_url: request.member_info.avatar_url.as_ref().map(|url| url.to_string()),
+            reason: request.member_info.reason.clone(),
+            is_seen: request.is_seen,
+            actions: Arc::new(RequestToJoinActions { inner: request }),
+        }
+    }
+}
+
+/// A listener for receiving new requests to a join a room.
+#[matrix_sdk_ffi_macros::export(callback_interface)]
+pub trait RequestsToJoinListener: Send + Sync {
+    fn call(&self, requests_to_join: Vec<RequestToJoin>);
+}
+
+/// An FFI representation of a request to join a room.
+#[derive(Debug, Clone, uniffi::Record)]
+pub struct RequestToJoin {
+    /// The event id of the event that contains the `knock` membership change.
+    pub event_id: String,
+    /// The user id of the user who's requesting to join the room.
+    pub user_id: String,
+    /// The room id of the room whose access was requested.
+    pub room_id: String,
+    /// The optional display name of the user who's requesting to join the room.
+    pub display_name: Option<String>,
+    /// The optional avatar url of the user who's requesting to join the room.
+    pub avatar_url: Option<String>,
+    /// An optional reason why the user wants join the room.
+    pub reason: Option<String>,
+    /// Whether the request to join has been marked as `seen` so it can be
+    /// filtered by the client.
+    pub is_seen: bool,
+    /// A set of actions to perform for this request to join.
+    pub actions: Arc<RequestToJoinActions>,
+}
+
+/// A set of actions to perform for a request to join.
+#[derive(Debug, Clone, uniffi::Object)]
+pub struct RequestToJoinActions {
+    inner: matrix_sdk::room::request_to_join::RequestToJoinRoom,
+}
+
+#[matrix_sdk_ffi_macros::export]
+impl RequestToJoinActions {
+    /// Accepts the request to join by inviting the user to the room.
+    pub async fn accept(&self) -> Result<(), ClientError> {
+        self.inner.accept().await.map_err(Into::into)
+    }
+
+    /// Declines the request to join by kicking the user from the room with an
+    /// optional reason.
+    pub async fn decline(&self, reason: Option<String>) -> Result<(), ClientError> {
+        self.inner.decline(reason.as_deref()).await.map_err(Into::into)
+    }
+
+    /// Declines the request to join by banning the user from the room with an
+    /// optional reason.
+    pub async fn decline_and_ban(&self, reason: Option<String>) -> Result<(), ClientError> {
```

Comment:
maybe we want a ban action, and let the user chain? in the case the user doesn't have the right for both

Reply:
You mean kick, then ban? Can we ban a user who already left the room? 

Also, I'd expect the UI to not display the component that triggers this action rather than controlling trying to workaround the issue here.

Reply:
But then that's what we'll do. ` decline` just kicks the user from the room, `decline_and_ban` also [just bans them](https://github.com/matrix-org/matrix-rust-sdk/pull/4338/files#diff-fa0c59c707536a04b3c50458f03c0e9da8fe5bd67536b2655d0e5bb3adc0a0cfR59), we don't do anything else.